File: sdk/python/examples/langgraph/06_conditional_routing.py
# Copyright (c) 2025 Agentspan
# Licensed under the MIT License. See LICENSE file in the project root for details.

"""Conditional Routing — create_agent with sentiment-based routing tools.

Demonstrates:
    - Using create_agent with a system prompt that directs conditional routing behaviour
    - Tools for classifying sentiment and generating tailored responses
    - The LLM orchestrates routing logic server-side via tool calls

Requirements:
    - AGENTSPAN_SERVER_URL=http://localhost:8080/api
    - OPENAI_API_KEY for ChatOpenAI
"""

from langchain_core.tools import tool
from langchain_openai import ChatOpenAI
from agentspan.agents.langchain import create_agent
from agentspan.agents import AgentRuntime

llm = ChatOpenAI(model="gpt-4o-mini", temperature=0)


@tool
def respond_to_positive(text: str) -> str:
    """Generate an enthusiastic, warm reply for a positive message.

    Args:
        text: The positive message to respond to.
    """
    return (
        f"That's wonderful to hear! Your positive energy is contagious. "
        f"Keep up the great momentum — you're doing amazing things!"
    )


@tool
def respond_to_negative(text: str) -> str:
    """Generate an empathetic, supportive reply for a negative or distressing message.

    Args:
        text: The negative message to respond to.
    """
    return (
        f"I hear you, and I'm sorry you're going through this. "
        f"Your feelings are completely valid. Remember that difficult times are temporary, "
        f"and there are people who care about you and want to help."
    )


@tool
def respond_to_neutral(text: str) -> str:
    """Generate an informative, balanced reply for a neutral message.

    Args:
        text: The neutral message to respond to.
    """
    return (
        f"Thank you for sharing that. I'd be happy to help you explore this topic further "
        f"or provide additional context if you have specific questions."
    )


ROUTER_SYSTEM = """You are a sentiment-aware assistant.
For every message:
1. Classify the sentiment (positive, negative, or neutral)
2. Based on the sentiment, call the matching response tool:
   - Positive sentiment → respond_to_positive
   - Negative sentiment → respond_to_negative
   - Neutral sentiment → respond_to_neutral
3. Return the tool's response as your answer.
"""

graph = create_agent(
    llm,
    tools=[respond_to_positive, respond_to_negative, respond_to_neutral],
    name="sentiment_router",
    system_prompt=ROUTER_SYSTEM,
)

if __name__ == "__main__":
    with AgentRuntime() as runtime:
        result = runtime.run(graph, "I just got promoted at work and I'm thrilled!")
        print(f"Status: {result.status}")
        result.print_result()

File: sdk/typescript/examples/langgraph/06-conditional-routing.ts
/**
 * Conditional Routing -- StateGraph with addConditionalEdges.
 *
 * Demonstrates:
 *   - Using addConditionalEdges to branch based on state content
 *   - A sentiment classifier that routes to positive, negative, or neutral handlers
 *   - Multiple terminal nodes converging to END
 */

import { StateGraph, START, END, Annotation } from '@langchain/langgraph';
import { AgentRuntime } from '../../src/index.js';

// ---------------------------------------------------------------------------
// State schema
// ---------------------------------------------------------------------------
const SentimentState = Annotation.Root({
  text: Annotation<string>({
    reducer: (_prev: string, next: string) => next ?? _prev,
    default: () => '',
  }),
  sentiment: Annotation<string>({
    reducer: (_prev: string, next: string) => next ?? _prev,
    default: () => 'neutral',
  }),
  response: Annotation<string>({
    reducer: (_prev: string, next: string) => next ?? _prev,
    default: () => '',
  }),
});

type State = typeof SentimentState.State;

// ---------------------------------------------------------------------------
// Node functions
// ---------------------------------------------------------------------------
function classify(state: State): Partial<State> {
  const text = state.text.toLowerCase();
  const positiveWords = ['thrilled', 'promoted', 'love', 'great', 'happy', 'amazing', 'excellent'];
  const negativeWords = ['sad', 'angry', 'terrible', 'hate', 'awful', 'disappointed', 'frustrated'];

  const posCount = positiveWords.filter((w) => text.includes(w)).length;
  const negCount = negativeWords.filter((w) => text.includes(w)).length;

  if (posCount > negCount) return { sentiment: 'positive' };
  if (negCount > posCount) return { sentiment: 'negative' };
  return { sentiment: 'neutral' };
}

function routeSentiment(state: State): string {
  return state.sentiment;
}

function handlePositive(_state: State): Partial<State> {
  return {
    response:
      "That's wonderful news! Congratulations on your promotion! " +
      'Your hard work and dedication are clearly paying off. Keep up the amazing work!',
  };
}

function handleNegative(_state: State): Partial<State> {
  return {
    response:
      "I'm sorry to hear that. It's completely valid to feel that way. " +
      'Remember, difficult times are temporary and things can improve. ' +
      'Is there anything specific I can help with?',
  };
}

function handleNeutral(_state: State): Partial<State> {
  return {
    response:
      "Thank you for sharing that. I'm here to help if you need anything. " +
      "Feel free to ask me any questions or share more about what's on your mind.",
  };
}

// ---------------------------------------------------------------------------
// Build the graph
// ---------------------------------------------------------------------------
const builder = new StateGraph(SentimentState);
builder.addNode('classify', classify);
builder.addNode('positive', handlePositive);
builder.addNode('negative', handleNegative);
builder.addNode('neutral', handleNeutral);
builder.addEdge(START, 'classify');
builder.addConditionalEdges('classify', routeSentiment, {
  positive: 'positive',
  negative: 'negative',
  neutral: 'neutral',
});
builder.addEdge('positive', END);
builder.addEdge('negative', END);
builder.addEdge('neutral', END);

const graph = builder.compile();

// Add agentspan metadata for extraction (no LLM in this pipeline example)
(graph as any)._agentspan = {
  model: 'openai/gpt-4o-mini',
  tools: [],
  framework: 'langgraph',
};

// ---------------------------------------------------------------------------
// Run on agentspan
// ---------------------------------------------------------------------------
async function main() {
  const runtime = new AgentRuntime();
  try {
    // Deploy to server. CLI alternative (recommended for CI/CD):
    //   agentspan deploy <module>
    await runtime.deploy(graph);
    await runtime.serve(graph);

    // Quick test: uncomment below (and comment out serve) to run directly.
    // const result = await runtime.run(
    // graph,
    // "I just got promoted at work and I'm thrilled!",
    // );
    // console.log('Status:', result.status);
    // result.printResult();
  } finally {
    await runtime.shutdown();
  }
}

// Only run when executed directly (not when imported for discovery)
if (process.argv[1]?.endsWith('06-conditional-routing.ts') || process.argv[1]?.endsWith('06-conditional-routing.js')) {
  main().catch(console.error);
}
